core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -70,35 +88,164 @@
   private volatile long mLastPrimaryStartSequenceNumber = 0;
   private volatile long mNextSequenceNumberToRead = 0;
   private volatile boolean mSnapshotting = false;
+  private volatile boolean mIsLeader = false;
+
   // The start time of the most recent snapshot
   private volatile long mLastSnapshotStartTime = 0;
   /** Used to control applying to masters. */
   private BufferedJournalApplier mJournalApplier;
+  private final SimpleStateMachineStorage mStorage = new SimpleStateMachineStorage();
+  private RaftGroupId mRaftGroupId;
 
   /**
-   * @param journals master journals; these journals are still owned by the caller, not by the
-   *        journal state machine
-   * @param journalSinks a supplier for journal sinks
+   * @param journals     master journals; these journals are still owned by the caller, not by the
+   *                     journal state machine
+   * @param journalSystem the raft journal system
    */
-  public JournalStateMachine(Map<String, RaftJournal> journals,
-      Supplier<Set<JournalSink>> journalSinks) {
+  public JournalStateMachine(Map<String, RaftJournal> journals, RaftJournalSystem journalSystem) {
     mJournals = journals;
-    mJournalApplier = new BufferedJournalApplier(journals, journalSinks);
+    mJournalApplier = new BufferedJournalApplier(journals,
+        () -> journalSystem.getJournalSinks(null));
     resetState();
     LOG.info("Initialized new journal state machine");
+    mJournalSystem = journalSystem;
+  }
+
+  private static <T> CompletableFuture<T> completeExceptionally(Exception e) {
+    final CompletableFuture<T> future = new CompletableFuture<>();
+    future.completeExceptionally(e);
+    return future;
+  }
+
+  @Override
+  public void initialize(RaftServer server, RaftGroupId groupId,
+      RaftStorage raftStorage) throws IOException {
+    getLifeCycle().startAndTransition(() -> {
+      super.initialize(server, groupId, raftStorage);
+      mRaftGroupId = groupId;
+      mStorage.init(raftStorage);
+      loadSnapshot(mStorage.getLatestSnapshot());
+    });
+  }
+
+  @Override
+  public void reinitialize() throws IOException {
```

Comment:
I'm curious. Why does `reinitialize()` and `initialize()` look completely different? I would have imagined that reinitialize would be quite similar to initialize, or would share some code?

Reply:
initialize() is invoked when the server is first started. reinitialize() is invoked when a follower got a new snapshot from leader and need to reload. 

There isn't much difference except maybe two things:
- the `initialize()` does a little more to initialize the group id and raft storage.
- `mStorage.loadLatestSnapshot()` was implicitly called in `initialize()` as part of `mStorage.init(raftStorage);`